shell/common/gl_context_switch.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_COMMON_GL_CONTEXT_SWITCH_H_
+#define FLUTTER_SHELL_COMMON_GL_CONTEXT_SWITCH_H_
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/macros.h"
+
+namespace flutter {
+
+//------------------------------------------------------------------------------
+/// An abstract class represents a gl context that can be switched by
+/// GLContextSwitch
+///
+/// The subclass should wrap a "Context" object inside this class. For example,
+/// in iOS while using GL rendering surface, the subclass should wrap an
+/// |EAGLContext|.
+class SwitchableGLContext {
+ public:
+  //------------------------------------------------------------------------------
+  /// Implement this to set the context wrapped by this |SwitchableGLContext|
+  /// object to the current context.
+  virtual bool SetCurrent() = 0;
+
+  //------------------------------------------------------------------------------
+  /// Implement this to remove the context wrapped by this |SwitchableGLContext|
+  /// object from current context;
+  virtual bool RemoveCurrent() = 0;
+
+  SwitchableGLContext();
+
+  virtual ~SwitchableGLContext();
+
+  FML_DISALLOW_COPY_AND_ASSIGN(SwitchableGLContext);
+};
+
+class GLContextResult {
```

Comment:
Why have this class instead of using a bool?

Reply:
We need to use `GLContextSwitch` on iOS to prevent plugins to pollute our context. For methods like `MakeFlutterContextCurrent`, we need to return a switch instead of a bool. But for platforms that doesn't require the switch (iOS software for example), we still need to return a same object. Making them returning a `switch` doesn't make sense as they are not actually switching context. So I thought to make this base class so at least the platforms that doesn't require a switch can return a same class.